src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -122,16 +122,25 @@ internal class TokenAcquisition : ITokenAcquisition, ITokenAcquisitionInternal
 
             try
             {
+                string? userFlow = context.Principal?.Claims?.FirstOrDefault(c => c.Type == ClaimConstants.UserFlow)?.Value;
                 _application = await GetOrBuildConfidentialClientApplicationAsync().ConfigureAwait(false);
 
                 // Do not share the access token with ASP.NET Core otherwise ASP.NET will cache it and will not send the OAuth 2.0 request in
                 // case a further call to AcquireTokenByAuthorizationCodeAsync in the future is required for incremental consent (getting a code requesting more scopes)
                 // Share the ID Token though
-                var result = await _application
-                    .AcquireTokenByAuthorizationCode(scopes.Except(_scopesRequestedByMsal), context.ProtocolMessage.Code)
-                    .WithSendX5C(_microsoftIdentityOptions.SendX5C)
-                    .ExecuteAsync()
-                    .ConfigureAwait(false);
+                var builder = _application
+                    .AcquireTokenByAuthorizationCode(scopes.Except(_scopesRequestedByMsal), context.ProtocolMessage.Code);
+
+                if (_microsoftIdentityOptions.IsB2C)
+                {
+                    var authority = $"{_applicationOptions.Instance}{Constants.Tfp}/{_microsoftIdentityOptions.Domain}/{userFlow ?? _microsoftIdentityOptions.DefaultUserFlow}";
+                    builder.WithB2CAuthority(authority);
+                }
+
+                var result = await builder.WithSendX5C(_microsoftIdentityOptions.SendX5C)
```

Comment:
Maybe move `WithSendX5C` up to chain off `AcquireTokenByAuthorizationCode`? So that build code is together and this line just executes?

Reply:
yes. good point. thx.